File: utils/data_processor.py
"""
Data Processing Module

This module implements SOLID principles in the following ways:

Single Responsibility Principle (SRP):
- Each function has a single, well-defined purpose
- Statistical calculations are separated from data processing
- Data cleaning and transformation are handled separately

Open/Closed Principle (OCP):
- New statistical measures can be added without modifying existing ones
- Data processing pipelines are extensible
- New data types can be supported without changing core functionality

Liskov Substitution Principle (LSP):
- All statistical functions follow the same interface pattern
- Data processing functions maintain consistent behavior
- Type checking ensures proper substitution

Interface Segregation Principle (ISP):
- Statistical functions are grouped by type
- Data processing functions are separated by responsibility
- Helper functions are isolated by purpose

Dependency Inversion Principle (DIP):
- The module depends on abstractions (pandas DataFrame) rather than concrete implementations
- Statistical calculations are independent of data source
- Processing functions are decoupled from specific data formats
"""

import pandas as pd
import numpy as np
from scipy import stats
from typing import Dict, Any
import io
import logging
from utils.json_sanitizer import sanitize_json

logger = logging.getLogger(__name__)

def calculate_numeric_stats(series: pd.Series) -> Dict[str, float]:
    """Calculate comprehensive statistics for a numeric series."""
    clean_series = series.replace([np.inf, -np.inf], np.nan).dropna()
    
    if len(clean_series) < 2:
        return {
            'type': 'numeric',
            'mean': np.nan,
            'median': np.nan,
            'std': np.nan,
            'cv': np.nan,
            'kurtosis': np.nan,
            'skewness': np.nan,
            'min': np.nan,
            'max': np.nan,
            'iqr': np.nan,
            'q1': np.nan,
            'q3': np.nan
        }
    
    q1 = float(clean_series.quantile(0.25))
    q3 = float(clean_series.quantile(0.75))
    iqr = float(q3 - q1)
    mean = float(clean_series.mean())
    std = float(clean_series.std())
    
    # Calculate outlier bounds
    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr
    outliers = clean_series[(clean_series < lower_bound) | (clean_series > upper_bound)]
    
    return {
        'type': 'numeric',
        'mean': mean,
        'median': float(clean_series.median()),
        'std': std,
        'cv': float((std / mean * 100) if mean != 0 else np.nan),
        'kurtosis': float(stats.kurtosis(clean_series, fisher=True)),
        'skewness': float(stats.skew(clean_series)),
        'min': float(clean_series.min()),
        'max': float(clean_series.max()),
        'iqr': iqr,
        'q1': q1,
        'q3': q3,
        'outliers_count': len(outliers),
        'outliers_percentage': float(len(outliers) / len(clean_series) * 100)
    }

def calculate_data_quality(df: pd.DataFrame) -> Dict[str, Dict[str, Any]]:
    """Calculate comprehensive data quality metrics."""
    
    # Calculate completeness (percentage of non-null values)
    completeness = float((1 - df.isnull().sum().sum() / (df.shape[0] * df.shape[1])) * 100)
    
    # Calculate uniqueness (average percentage of unique values across columns)
    uniqueness = float(np.mean([len(df[col].unique()) / len(df) * 100 for col in df.columns]))
    
    # Calculate consistency (percentage of values within expected ranges/patterns)
    numeric_cols = df.select_dtypes(include=[np.number]).columns
    consistency_scores = []
    
    for col in numeric_cols:
        clean_series = df[col].replace([np.inf, -np.inf], np.nan).dropna()
        if len(clean_series) > 0:
            # Calculate z-scores
            z_scores = np.abs(stats.zscore(clean_series))
            # Consider values within 3 standard deviations as consistent
            consistency_scores.append(np.mean(z_scores <= 3) * 100)
    
    consistency = float(np.mean(consistency_scores)) if consistency_scores else 0.0
    
    return {
        'completeness': {
            'score': f"{completeness:.1f}%",
            'rating': 'Excellent' if completeness >= 95 else 'Good' if completeness >= 80 else 'Poor'
        },
        'uniqueness': {
            'score': f"{uniqueness:.1f}%",
            'rating': 'Excellent' if uniqueness >= 80 else 'Good' if uniqueness >= 50 else 'Poor'
        },
        'consistency': {
            'score': f"{consistency:.1f}%",
            'rating': 'Excellent' if consistency >= 90 else 'Good' if consistency >= 70 else 'Poor'
        }
    }

def process_data(df: pd.DataFrame) -> Dict[str, Any]:
    """Process uploaded data and generate statistics."""
    
    df = sanitize_json(df)
    
    try:
        # Clean column names and handle potential single-column CSV issue
        if len(df.columns) == 1 and ',' in df.columns[0]:
            # The file was not properly parsed, try to parse it again
            first_col_name = df.columns[0]
            if isinstance(df.iloc[0, 0], str) and ',' in df.iloc[0, 0]:
                # Convert the single column to a string and split by newlines
                data_str = '\n'.join([first_col_name] + df[first_col_name].astype(str).tolist())
                # Read the string as a CSV
                df = pd.read_csv(io.StringIO(data_str))

        # Clean column names
        df.columns = df.columns.str.strip()
        
        numeric_columns = df.select_dtypes(include=[np.number]).columns
        categorical_columns = df.select_dtypes(include=['object']).columns
        
        # Calculate data quality metrics
        quality_metrics = calculate_data_quality(df)
        
        stats = {
            'summary': {
                'rows': len(df),
                'columns': len(df.columns),
                'numeric_columns': len(numeric_columns),
                'categorical_columns': len(categorical_columns),
                'memory_usage': f"{df.memory_usage(deep=True).sum() / 1024 / 1024:.2f} MB"
            },
            'data_quality': quality_metrics,
            'column_stats': {},
            'preview': df.head(5).to_dict('records'),
            'columns': list(df.columns)
        }
        
        # Calculate statistics for numeric columns
        for col in numeric_columns:
            col_stats = calculate_numeric_stats(df[col])
            stats['column_stats'][col] = {
                'statistics': col_stats,
                'key_insights': generate_column_insights(df[col], col_stats)
            }
        
        # Calculate statistics for categorical columns
        for col in categorical_columns:
            value_counts = df[col].value_counts()
            stats['column_stats'][col] = {
                'type': 'categorical',
                'unique_values': len(value_counts),
                'top_values': value_counts.head(5).to_dict()
            }
        
        return stats
    except Exception as e:
        logger.exception("Error processing data")
        return None

def generate_column_insights(series: pd.Series, stats: Dict[str, float]) -> list:
    """Generate detailed insights about the column based on its statistics."""
    insights = []
    
    # Check for data distribution
    if abs(stats['mean'] - stats['median']) > stats['std'] * 0.1:
        if stats['mean'] > stats['median']:
            insights.append(f"Data shows positive skew with mean ({stats['mean']:.2f}) greater than median ({stats['median']:.2f})")
        else:
            insights.append(f"Data shows negative skew with mean ({stats['mean']:.2f}) less than median ({stats['median']:.2f})")
    
    # Check for skewness
    if abs(stats['skewness']) > 1:
        direction = "positive" if stats['skewness'] > 0 else "negative"
        severity = "highly" if abs(stats['skewness']) > 2 else "moderately"
        insights.append(f"{severity.capitalize()} {direction} skewed distribution (skewness: {stats['skewness']:.2f})")
    
    # Check for kurtosis
    if abs(stats['kurtosis']) > 2:
        if stats['kurtosis'] > 0:
            insights.append(f"Heavy-tailed distribution with more extreme values than normal (kurtosis: {stats['kurtosis']:.2f})")
        else:
            insights.append(f"Light-tailed distribution with fewer extreme values than normal (kurtosis: {stats['kurtosis']:.2f})")
    
    # Check for variability
    if stats['cv'] > 100:
        severity = "extremely" if stats['cv'] > 150 else "highly"
        insights.append(f"{severity.capitalize()} variable data (CV: {stats['cv']:.1f}%), indicating wide value dispersion")
    
    # Check for outliers
    if 'outliers_count' in stats and stats['outliers_count'] > 0:
        insights.append(f"Found {stats['outliers_count']} potential outliers ({stats['outliers_percentage']:.1f}% of values)")
    
    # Check for data range
    range_size = stats['max'] - stats['min']
    if range_size > 0:
        insights.append(f"Data spans from {stats['min']:.2f} to {stats['max']:.2f} (range: {range_size:.2f})")
    
    # Check for concentration
    if 'iqr' in stats and stats['iqr'] > 0:
        central_range = f"50% of values fall between {stats['q1']:.2f} and {stats['q3']:.2f} (IQR: {stats['iqr']:.2f})"
        insights.append(central_range)
    
    return insights

def chunk_process_data(df: pd.DataFrame, chunk_size: int = 10000) -> Dict[str, Any]:
    """Process large datasets in chunks to avoid memory issues."""
    
    total_rows = len(df)
    chunks = []
    chunk_stats = []
    
    # Process data in chunks
    for i in range(0, total_rows, chunk_size):
        chunk = df.iloc[i:i + chunk_size]
        chunk_stats.append(process_data(chunk))
    
    # Combine chunk statistics
    combined_stats = {
        'summary': {
            'rows': total_rows,
            'columns': len(df.columns),
            'numeric_columns': len(df.select_dtypes(include=[np.number]).columns),
            'categorical_columns': len(df.select_dtypes(include=['object']).columns),
            'memory_usage': f"{df.memory_usage(deep=True).sum() / 1024 / 1024:.2f} MB",
            'chunks_processed': len(chunk_stats)
        },
        'column_stats': {},
        'preview': df.head(5).to_dict('records'),
        'columns': list(df.columns)
    }
    
    # Merge column statistics from all chunks
    numeric_columns = df.select_dtypes(include=[np.number]).columns
    for col in numeric_columns:
        combined_stats['column_stats'][col] = {
            'statistics': calculate_numeric_stats(df[col]),
            'key_insights': generate_column_insights(df[col], calculate_numeric_stats(df[col]))
        }
    
    # Handle categorical columns
    categorical_columns = df.select_dtypes(include=['object']).columns
    for col in categorical_columns:
        value_counts = df[col].value_counts()
        combined_stats['column_stats'][col] = {
            'type': 'categorical',
            'unique_values': len(value_counts),
            'top_values': value_counts.head(5).to_dict()
        }
    
    return combined_stats

"""
Statistical Processing Layer

SRP: Handles only statistical calculations
OCP: New statistical methods can be added without modification
LSP: All statistical functions follow same interface
"""

"""
Data Transformation Layer

SRP: Responsible only for data transformation
ISP: Transformations are separated by data type
DIP: Independent of specific data formats
"""

"""
Data Quality Analysis Layer

SRP: Handles only data quality assessment
OCP: New quality metrics can be added without modification
ISP: Quality checks are separated by type
"""

"""
Insight Generation Layer

SRP: Responsible only for generating insights
OCP: New insight types can be added without modification
DIP: Independent of specific data sources
"""

File: static/js/dataProcessor.js
/**
 * Data Processing Module
 * 
 * This module implements SOLID principles in the following ways:
 * 
 * Single Responsibility Principle (SRP):
 * - The module is solely responsible for data processing and transformation
 * - Each processing function handles a specific type of data transformation
 * - Statistical calculations are isolated from data structure manipulation
 * 
 * Open/Closed Principle (OCP):
 * - New data processing functions can be added without modifying existing ones
 * - Data transformation pipelines are extensible
 * - Statistical calculations can be extended for new types of analysis
 * 
 * Liskov Substitution Principle (LSP):
 * - All data processing functions follow consistent input/output patterns
 * - Data transformation functions are interchangeable when they operate on the same data type
 * - Statistical calculations maintain consistent behavior across different data types
 * 
 * Interface Segregation Principle (ISP):
 * - Data processing functions are grouped by data type (categorical, numerical, etc.)
 * - Statistical calculations are separated based on their specific purposes
 * - Helper functions are organized by their specific responsibilities
 * 
 * Dependency Inversion Principle (DIP):
 * - The module depends on data abstractions rather than concrete implementations
 * - Processing functions are independent of specific data sources
 * - Statistical calculations are decoupled from specific data structures
 */

function processData(data) {
    try {
        console.log('Processing data input:', data);
        let processedData = {};

        // Ensure we have valid data
        if (!data || !data.preview || !Array.isArray(data.preview) || data.preview.length === 0) {
            console.warn('Invalid or empty data provided to processData');
            return null;
        }

        // Extract column information
        const columns = Object.keys(data.preview[0] || {});
        console.log('Found columns:', columns);

        if (columns.length === 0) {
            console.warn('No columns found in data');
            return null;
        }

        // Calculate frequencies and statistics for all columns
        const columnStats = {};
        columns.forEach(column => {
            console.log(`Processing column: ${column}`);
            const values = data.preview.map(row => row[column]);
            console.log(`Values for ${column}:`, values);
            
            const validValues = values.filter(v => v != null && v !== '');
            console.log(`Valid values for ${column}:`, validValues);
            
            if (validValues.length === 0) {
                console.log(`Skipping empty column: ${column}`);
                return;
            }

            // Determine column type
            const isNumeric = validValues.every(v => !isNaN(v) && typeof v !== 'boolean');
            const uniqueValues = new Set(validValues);
            const isCategorical = uniqueValues.size <= Math.min(20, validValues.length / 2);
            
            console.log(`Column ${column} type:`, isNumeric ? 'numeric' : 'categorical');
            console.log(`Column ${column} unique values:`, Array.from(uniqueValues));

            // Calculate frequencies
            const frequencies = {};
            validValues.forEach(value => {
                frequencies[value] = (frequencies[value] || 0) + 1;
            });
            
            console.log(`Frequencies for ${column}:`, frequencies);

            columnStats[column] = {
                type: isNumeric ? 'numeric' : 'categorical',
                frequencies,
                uniqueValues: Array.from(uniqueValues),
                count: validValues.length
            };

            // Add numeric statistics if applicable
            if (isNumeric) {
                const numericValues = validValues.map(v => Number(v));
                const sorted = [...numericValues].sort((a, b) => a - b);
                columnStats[column] = {
                    ...columnStats[column],
                    min: Math.min(...numericValues),
                    max: Math.max(...numericValues),
                    mean: numericValues.reduce((a, b) => a + b, 0) / numericValues.length,
                    median: sorted[Math.floor(sorted.length / 2)],
                    q1: sorted[Math.floor(sorted.length * 0.25)],
                    q3: sorted[Math.floor(sorted.length * 0.75)]
                };
            }
        });

        console.log('Column stats:', columnStats);

        // Process categorical data
        const categoricalColumns = Object.entries(columnStats)
            .filter(([_, stats]) => stats.type === 'categorical')
            .map(([col]) => col);
            
        console.log('Categorical columns:', categoricalColumns);

        if (categoricalColumns.length > 0) {
            processedData.categorical = {};
            categoricalColumns.forEach(col => {
                console.log(`Processing categorical column: ${col}`);
                const stats = columnStats[col];
                console.log(`Stats for ${col}:`, stats);
                
                if (stats && stats.frequencies) {
                    processedData.categorical[col] = {
                        frequencies: stats.frequencies,
                        total: stats.count,
                        unique_values: stats.uniqueValues
                    };
                    console.log(`Processed categorical data for ${col}:`, processedData.categorical[col]);
                } else {
                    console.warn(`Missing stats or frequencies for column: ${col}`);
                }
            });
        }

        // Process numeric data
        const numericColumns = Object.entries(columnStats)
            .filter(([_, stats]) => stats.type === 'numeric')
            .map(([col]) => col);

        if (numericColumns.length > 0) {
            const numericData = {};
            
            // Prepare histogram data
            if (numericColumns[0]) {
                const histogramData = prepareHistogramData(data, numericColumns[0], columnStats);
                if (histogramData) {
                    numericData.histogram = histogramData;
                }
            }
            
            // Prepare scatter data
            if (numericColumns.length >= 2) {
                const scatterData = prepareScatterData(data, numericColumns[0], numericColumns[1]);
                if (scatterData) {
                    numericData.scatter = scatterData;
                }
            }
            
            // Prepare boxplot data
            const boxplotData = prepareBoxplotData(data, numericColumns, columnStats);
            if (boxplotData) {
                numericData.boxplot = boxplotData;
            }
            
            // Prepare heatmap data
            if (numericColumns.length >= 2) {
                const heatmapData = prepareHeatmapData(data, numericColumns);
                if (heatmapData) {
                    numericData.heatmap = heatmapData;
                }
            }

            if (Object.keys(numericData).length > 0) {
                processedData = { ...processedData, ...numericData };
            }
        }

        // Add column statistics
        processedData.column_stats = columnStats;
        
        // Add preview data
        processedData.preview = data.preview;

        console.log('Final processed data:', processedData);
        return processedData;
    } catch (error) {
        console.error('Error in processData:', error);
        return null;
    }
}

/**
 * Data Validation Layer
 * 
 * SRP: Responsible only for validating input data
 * ISP: Validation rules are separated by data type
 * DIP: Validation is independent of data processing implementation
 */

function validateNumericValues(values) {
    return values.filter(v => 
        v !== null && 
        v !== undefined && 
        !isNaN(v) && 
        typeof v !== 'boolean'
    ).map(v => Number(v));
}

/**
 * Statistical Processing Layer
 * 
 * SRP: Handles only statistical calculations
 * OCP: New statistical methods can be added without changing existing ones
 * ISP: Statistical functions are grouped by type of analysis
 */

function prepareHistogramData(data, column, columnStats) {
    if (!column || !columnStats[column]) return null;

    try {
        const stats = columnStats[column];
        const values = data.preview
            .map(row => row[column])
            .filter(v => v != null && !isNaN(v))
            .map(Number);

        if (values.length === 0) return null;

        // Create histogram bins
        const binCount = Math.min(10, Math.ceil(Math.sqrt(values.length)));
        const min = stats.min;
        const max = stats.max;
        const binWidth = (max - min) / binCount;
        const bins = Array(binCount).fill(0);

        values.forEach(value => {
            const binIndex = Math.min(
                Math.floor((value - min) / binWidth),
                binCount - 1
            );
            bins[binIndex]++;
        });

        return {
            column,
            values: bins,
            bins: Array(binCount).fill(0).map((_, i) => 
                `${(min + i * binWidth).toFixed(2)} - ${(min + (i + 1) * binWidth).toFixed(2)}`
            )
        };
    } catch (error) {
        console.error('Error preparing histogram data:', error);
        return null;
    }
}

function prepareScatterData(data, columnX, columnY) {
    if (!columnX || !columnY || !data.preview || !data.preview.length) {
        return null;
    }

    try {
        const pairedData = data.preview
            .map(row => [row[columnX], row[columnY]])
            .filter(([x, y]) => 
                x !== null && y !== null && 
                !isNaN(x) && !isNaN(y)
            );

        if (!pairedData.length) return null;

        return {
            x: pairedData.map(d => d[0]),
            y: pairedData.map(d => d[1]),
            xLabel: columnX,
            yLabel: columnY
        };
    } catch (error) {
        console.error('Error preparing scatter data:', error);
        return null;
    }
}

function prepareBoxplotData(data, columns, columnStats) {
    if (!columns || !columns.length) {
        return null;
    }

    try {
        return columns
            .filter(col => columnStats[col] && columnStats[col].type === 'numeric')
            .map(col => ({
                name: col,
                stats: {
                    min: columnStats[col].min,
                    q1: columnStats[col].q1 || columnStats[col].min,
                    median: columnStats[col].median,
                    q3: columnStats[col].q3 || columnStats[col].max,
                    max: columnStats[col].max
                }
            }));
    } catch (error) {
        console.error('Error preparing boxplot data:', error);
        return null;
    }
}

function prepareHeatmapData(data, columns) {
    if (!columns || columns.length < 2) {
        return null;
    }

    try {
        const correlationMatrix = {
            columns: columns,
            values: []
        };

        // Calculate correlations between numeric columns
        for (let i = 0; i < columns.length; i++) {
            for (let j = 0; j < columns.length; j++) {
                const correlation = i === j ? 1 : calculateCorrelation(
                    data.preview.map(row => row[columns[i]]),
                    data.preview.map(row => row[columns[j]])
                );
                correlationMatrix.values.push([i, j, correlation || 0]);
            }
        }

        return correlationMatrix;
    } catch (error) {
        console.error('Error preparing heatmap data:', error);
        return null;
    }
}

function calculateCorrelation(x, y) {
    try {
        const validPairs = x.map((val, i) => [val, y[i]])
            .filter(([a, b]) => 
                a !== null && b !== null && 
                !isNaN(a) && !isNaN(b)
            );

        if (validPairs.length < 2) return null;

        const n = validPairs.length;
        const [xs, ys] = validPairs.reduce(
            ([accX, accY], [currX, currY]) => [
                [...accX, Number(currX)],
                [...accY, Number(currY)]
            ],
            [[], []]
        );

        const sumX = xs.reduce((a, b) => a + b, 0);
        const sumY = ys.reduce((a, b) => a + b, 0);
        const sumXY = xs.reduce((sum, x, i) => sum + x * ys[i], 0);
        const sumX2 = xs.reduce((sum, x) => sum + x * x, 0);
        const sumY2 = ys.reduce((sum, y) => sum + y * y, 0);

        const numerator = n * sumXY - sumX * sumY;
        const denominator = Math.sqrt(
            (n * sumX2 - sumX * sumX) * (n * sumY2 - sumY * sumY)
        );

        return denominator === 0 ? 0 : numerator / denominator;
    } catch (error) {
        console.error('Error calculating correlation:', error);
        return null;
    }
}

function prepareCategoricalData(data, columns) {
    if (!columns || !columns.length || !data.preview || !data.preview.length) {
        return null;
    }

    try {
        const result = {};
        
        for (const column of columns) {
            // Count frequency of each category
            const frequencies = {};
            data.preview.forEach(row => {
                const value = row[column];
                if (value !== null && value !== undefined) {
                    frequencies[value] = (frequencies[value] || 0) + 1;
                }
            });

            // Only include if we have valid frequencies
            if (Object.keys(frequencies).length > 0) {
                // Sort categories by frequency
                const sortedCategories = Object.entries(frequencies)
                    .sort(([,a], [,b]) => b - a)
                    .reduce((obj, [key, value]) => {
                        obj[key] = value;
                        return obj;
                    }, {});

                result[column] = {
                    frequencies: sortedCategories,
                    total: Object.values(frequencies).reduce((a, b) => a + b, 0)
                };
            }
        }

        return Object.keys(result).length > 0 ? result : null;
    } catch (error) {
        console.error('Error preparing categorical data:', error);
        return null;
    }
}

// Add new function to handle education vs income relationship
function prepareEducationIncomeData(data, educationColumn, incomeColumn) {
    if (!data.preview || !data.preview.length) {
        return null;
    }

    try {
        // Group data by education level
        const educationGroups = {};
        
        data.preview.forEach(row => {
            const education = row[educationColumn];
            const income = parseFloat(row[incomeColumn]);
            
            if (education && !isNaN(income)) {
                if (!educationGroups[education]) {
                    educationGroups[education] = [];
                }
                educationGroups[education].push(income);
            }
        });

        // Calculate statistics for each education level
        const result = Object.entries(educationGroups).map(([education, incomes]) => {
            const avg = incomes.reduce((a, b) => a + b, 0) / incomes.length;
            const sorted = [...incomes].sort((a, b) => a - b);
            const median = sorted[Math.floor(sorted.length / 2)];
            
            return {
                education,
                averageIncome: avg,
                medianIncome: median,
                count: incomes.length,
                min: Math.min(...incomes),
                max: Math.max(...incomes)
            };
        });

        return {
            educationLevels: result.map(r => r.education),
            averageIncomes: result.map(r => r.averageIncome),
            medianIncomes: result.map(r => r.medianIncome),
            counts: result.map(r => r.count),
            ranges: result.map(r => ({ min: r.min, max: r.max }))
        };
    } catch (error) {
        console.error('Error preparing education vs income data:', error);
        return null;
    }
}

/**
 * Data Transformation Layer
 * 
 * SRP: Responsible only for data structure transformation
 * OCP: New transformation methods can be added without modification
 * LSP: All transformations maintain consistent data patterns
 */
